### PR TITLE
Support system role in hf data processing

### PR DESCRIPTION
**Pull request by ChingTsai**
# Description
- Supporting `system` role in hf data processing
- Add test case for Qwen in `sft_data_processing_test.py`
- Refactoring `sft_data_processing_test.py` to support testing multiple tokenizer
- Fix incorrect left shift in targets_segmentation in `shift_and_refine`
   - see more details [here](https://github.com/AI-Hypercomputer/maxtext/pull/2970/files#r2707486614) 

FIXES: b/471740714

# Tests
### Unit test
```
MAXTEXT_PKG_DIR=<xxx>/maxtext/src/MaxText pytest tests/sft_data_processing_test.py
```
### E2E sft
```
python3 -m MaxText.sft.sft_trainer \
   src/MaxText/configs/sft.yml \
    run_name=$(date +%Y-%m-%d-%H-%M-%S) \
    base_output_directory=<xxx> \
    model_name=qwen3-4b \
    load_parameters_path=<xxx>/0/items \
    tokenizer_path=Qwen/Qwen3-4B \
    steps=53 \
    profiler=xplane \
    hf_path=arrow \
    dataset_type=hf \
    train_split=train \
    hf_train_files=<xxx>.arrow \
    hf_eval_files=<xxx>.arrow \
    per_device_batch_size=16 \
    max_target_length=1024 \
    learning_rate=5e-6 \
    warmup_steps_fraction=0.05 \
   data_shuffle_seed=42 \
    gradient_clipping_threshold=1 \
    weight_dtype=bfloat16
```

#### logs
[unpatched gpaste](https://paste.googleplex.com/4921812016300032)
[patched gpaste](https://paste.googleplex.com/4715041083490304)

Step 1 loss drop from `6.100826 ` to `5.839514 `

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/input_pipeline/_input_pipeline_utils.py [hunk 1]
```diff
@@ -688,6 +695,7 @@ def shift_left(x, pad_id, axis=1):
 def shift_and_refine(x, ignored_ids, axis=1):
   """Shift inputs, set segmentation to 0 when target element is in ignored_ids if provided"""
   x["targets"] = shift_left(x["targets"], ignored_ids[0], axis=axis)
+  x["targets_segmentation"] = shift_left(x["targets_segmentation"], 0, axis=axis)
```
Comment (ChingTsai) [inline src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: For Llama 2, since token_id `0` is `unknown_id`, the segmentation for pad_ids (added during packing) is updated correctly in lines 699-700. However, for Qwen3, token_id `0` is `!`,  therefore, we need to apply an explicit left shift to the segmentation here.

Comment (SurbhiJainUSC) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: Looks like this shifting is model specific. How do we ensure that it works for other models?

**Comment by SurbhiJainUSC** [comment]
Please rebase to fix Documentation Build CI.

File: src/MaxText/input_pipeline/_input_pipeline_utils.py (hunk 1 above)
Comment (aireenmei) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: Not sure I understand here. target_segmentation is a mask used here (https://github.com/AI-Hypercomputer/maxtext/blob/f9b1af066049a1a99252af92e4ffb1939ac632ec/src/MaxText/train.py#L152). 0 means masked token, unrelated to the token_id `0`

Comment (ChingTsai) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: Hi @SurbhiJainUSC and @aireenmei,

This change fixes an existing shift-left operation, so it  applies only to models using the `shift=True` in their processing pipeline.

As for this fix there, Please correct me if I'm wrong: Is `shift_and_refine` supposed to 1) shift targets left by one position, append a pad_token to the right, 2) synchronize `targets_segmentation ` by identifying all `ignore_id` positions in the new `targets` ?
In the original implementation, `targets_segmentation` did not shift alongside `targets`. Consequently, the first padding token added during packing was incorrectly assigned a segmentation value of `1`.

As shown in the example below, we currently see `!XXXXX` instead of the expected six X tokens at the end of the sequence. This occurs because the 0 (pad_token) still has its segmentation value set to `1`.
The proposed changes are intended to correct this behavior. Please let me know if this is the intended logic.

Many Thanks!

```python
import transformers
import numpy as np
from MaxText.input_pipeline._input_pipeline_utils import shift_left
def prepare_x():
    return {
        'inputs': np.array([[151644, 872, 198, 7841, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 9217, 151645, 198, 0, 0, 0, 0]]),
        'inputs_segmentation': np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0]]),
        'inputs_position': np.array([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0, 0, 0, 0]]),
        'targets': np.array([[151643, 151643, 151643, 151643, 151643, 151643, 151644, 77091, 198, 151667, 271, 151668, 271, 9217, 151645, 198, 0, 0, 0, 0]]),
        'targets_segmentation': np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0]]),
        'targets_position': np.array([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0, 0, 0, 0]])
    }
qwen3_tok = transformers.AutoTokenizer.from_pretrained("Qwen/Qwen3-4B",
        add_bos_token=False,
        add_eos_token=False,
        legacy=False,)

qwen3_x = prepare_x()
print("Before Shift:" + repr(qwen3_tok.decode(np.where(qwen3_x["targets_segmentation"][0] > 0, qwen3_x["targets"][0], qwen3_tok.encode("X")[0]))))
shift_and_refine(qwen3_x, ignored_ids=[151643, None])
print("After Shift:" + repr(qwen3_tok.decode(np.where(qwen3_x["targets_segmentation"][0] > 0, qwen3_x["targets"][0], qwen3_tok.encode("X")[0]))))
```
```
Before Shift:'<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|im_start|>assistant\n<think>\n\n</think>\n\nanswer<|im_end|>\nXXXX'
After Shift:'XXXXX<|im_start|>assistant\n<think>\n\n</think>\n\nanswer<|im_end|>\n!XXXX'
```

Comment (aireenmei) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: Thanks for the example. I understand it now. I see the problem is more on the packing side, we should use the padding token defined by the tokenizer instead of using 0 to pad packed sequence. In this example, if packing use 151643 to pad "inputs" and "targets", then in the shift function, the step "2) synchronize targets_segmentation  by identifying all ignore_id positions in the new targets" would convert that last token's targets_segmentation to 0. But since this hf pipeline is using a legacy packing Op that doesn't support custom padding token, your change of shifting targets_segmentation is the easier fix for this. This will fix for all the models that use a tokenizer where token_id 0 maps to meaningful token, and won't impact when the tokenizer actually use 0 as pad_id.

Comment (ChingTsai) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py:698]: Got it, that makes sense. Thanks for clarifying!

**Review by SurbhiJainUSC — APPROVED** [review]
